### PR TITLE
[cmclient] Seperate vchan to read and write

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

**Review by Eabramov84 — APPROVED** [review]

File: src/cmclient/cmclient.hpp [hunk 1]
```diff
@@ -107,12 +111,16 @@ class CMClient : public aos::sm::launcher::InstanceStatusReceiverItf,
...
+    aos::Thread<>                                                mThreadReader = {};
     atomic_t                                                     mState = {};
-    vch_handle                                                   mSMVChanHandler;
+    bool                                                         mWriteFail = {};
+    vch_handle                                                   mSMVChanHandlerWriter;
+    vch_handle                                                   mSMVChanHandlerReader;
     servicemanager_v3_SMIncomingMessages                         mIncomingMessage;
     servicemanager_v3_SMOutgoingMessages                         mOutgoingMessage;
     aos::Mutex                                                   mMutex;
+    aos::ConditionalVariable                                     mWaitWriteFail {mMutex};
```
Comment (al1img) [inline src/cmclient/cmclient.hpp:123]: It is used not only for write fail indication but for wake up write thread. Consider renaming it accordingly: `mWriterCondVar` for example.

File: src/cmclient/cmclient.hpp [hunk 2]
```diff
@@ -107,12 +111,16 @@ class CMClient : public aos::sm::launcher::InstanceStatusReceiverItf,
     aos::sm::launcher::LauncherItf*                              mLauncher = {};
     ResourceManager*                                             mResourceManager = {};
     DownloadReceiverItf*                                         mDownloader = {};
-    aos::Thread<>                                                mThread = {};
+    aos::Thread<>                                                mThreadWriter = {};
+    aos::Thread<>                                                mThreadReader = {};
     atomic_t                                                     mState = {};
-    vch_handle                                                   mSMVChanHandler;
+    bool                                                         mWriteFail = {};
```
Comment (al1img) [inline src/cmclient/cmclient.hpp:117]: It is not good the some internal states are handled in different places. We have `mState` add additional field there such a `eWriteFail`. Also `eConnected` is not used anymore - remove it.

File: src/cmclient/cmclient.cpp [hunk 3]
```diff
@@ -172,10 +154,80 @@ void CMClient::ConnectToCM()
...
+                continue;
+            }
+
+            {
+                aos::LockGuard lock(mMutex);
+
+                mWaitWriteFail.Wait([this] { return mWriteFail || atomic_test_bit(&mState, eFinish); });
+
+                mWriteFail = false;
+
```
Comment (al1img) [inline src/cmclient/cmclient.cpp:187]: Could be test after `vch_close`

File: src/cmclient/cmclient.cpp [hunk 4]
```diff
@@ -316,21 +368,24 @@ aos::Error CMClient::SendPBMessageToVChan()
 
     err = SendBufferToVChan((uint8_t*)&header, sizeof(VChanMessageHeader));
     if (!err.IsNone()) {
+        NotifyWriteFail();
```
Comment (al1img) [inline src/cmclient/cmclient.cpp:371]: You can send it once in `SendBufferToVChan`

File: src/cmclient/cmclient.cpp [hunk 5]
```diff
@@ -39,49 +44,20 @@ aos::Error CMClient::Init(LauncherItf& launcher, ResourceManager& resourceManage
...
+    return RunReader();
 }
 
 aos::Error CMClient::InstancesRunStatus(const aos::Array<aos::InstanceStatus>& instances)
 {
     aos::LockGuard lock(mMutex);
 
-    if (!atomic_test_bit(&mState, eConnected)) {
-        return AOS_ERROR_WRAP(aos::ErrorEnum::eWrongState);
+    if (atomic_test_bit(&mState, eFinish)) {
```
Comment (al1img) [inline src/cmclient/cmclient.cpp:59]: Could we move this condition inside `SendBufferToVChan`?

File: Kconfig [hunk 6]
```diff
@@ -9,9 +9,14 @@ config AOS_DOMD_ID
...
-config AOS_SM_VCHAN_PATH
+config AOS_SM_VCHAN_TX_PATH
 	string "Path to the SM vchan"
-	default "/local/domain/1/tmp/vchan/sm"
+	default "/local/domain/1/tmp/vchan/sm_tx"
+
+config AOS_SM_VCHAN_RX_PATH
+	string "Path to the SM vchan"
+	default "/local/domain/1/tmp/vchan/sm_rx"
+
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply Kconfig:19]: Done

File: src/cmclient/cmclient.cpp [hunk 7]
```diff
@@ -158,12 +134,18 @@ aos::Error CMClient::SendImageContentRequest(const ImageContentRequest& request)
...
+void CMClient::ConnectToCM(vch_handle& vchanHandler, const aos::String& vchanPath)
 {
     while (true) {
-        LOG_DBG() << "Connecting to vchan: " << cSMVChanPath << ", on dom: " << cDomdID;
+        LOG_DBG() << "Connecting to vchan: " << vchanPath << ", on dom: " << cDomdID;
+
+        int ret = 0;
+
+        {
+            aos::LockGuard lock(mMutex);
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/cmclient/cmclient.cpp:145]: Done

File: src/cmclient/cmclient.hpp (hunk 1 above)
Comment (MykolaSolyanko) [reply src/cmclient/cmclient.hpp:123]: Done

File: src/cmclient/cmclient.hpp (hunk 2 above)
Comment (MykolaSolyanko) [reply src/cmclient/cmclient.hpp:117]: Done

File: src/cmclient/cmclient.cpp (hunk 3 above)
Comment (MykolaSolyanko) [reply src/cmclient/cmclient.cpp:187]: Remove this variable

File: src/cmclient/cmclient.cpp (hunk 4 above)
Comment (MykolaSolyanko) [reply src/cmclient/cmclient.cpp:371]: Done

File: src/cmclient/cmclient.cpp (hunk 5 above)
Comment (MykolaSolyanko) [reply src/cmclient/cmclient.cpp:59]: Agree, it is much better.

File: src/cmclient/cmclient.hpp [hunk 8]
```diff
@@ -79,20 +79,24 @@ class CMClient : public aos::sm::launcher::InstanceStatusReceiverItf,
...
+    static constexpr auto cSMVChanRxPath = CONFIG_AOS_SM_VCHAN_RX_PATH;
     static constexpr auto cNodeID = CONFIG_AOS_NODE_ID;
     static constexpr auto cNodeType = CONFIG_AOS_NODE_TYPE;
     static constexpr auto cNumCPUs = CONFIG_AOS_NUM_CPU;
     static constexpr auto cTotalRAM = CONFIG_AOS_TOTAL_RAM;
     static constexpr auto cPartitionSize = CONFIG_AOS_PARTITION_SIZE;
     static constexpr auto CReadDelayUSec = 50000;
 
-    enum State { eConnected = 0, eFinish = 1 };
+    enum State { eConnected = 0, eFail = 1, eFinish = 2 };
```
Comment (al1img) [inline src/cmclient/cmclient.hpp:91]: eConnected is not used.

**Review by al1img — APPROVED** [review]

**Pull request merged**